File: typescript/packages/core/src/commands/builtin/generic_bind/builders/du.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { pathRulesActive } from '../../../../context/session_context.ts'
import { isEacces, isMissingPath } from '../../../../utils/errors.ts'
import { mountKey, mountPrefixOf, rekey } from '../../../../utils/key_prefix.ts'
import type { Accessor } from '../../../../accessor/base.ts'
import type { IndexCacheStore } from '../../../../cache/index/store.ts'
import { IOResult } from '../../../../io/types.ts'
import { FileType, PathSpec } from '../../../../types.ts'
import {
  DEFAULT_MAX_DU_ENTRIES,
  type ComputeEntries,
  type ComputeSize,
  type DuEntries,
  runDu,
} from '../../generic/du.ts'
import { type Builder, type CommandIO, resolveGlobOf } from '../adapter.ts'
import { compareCodePoints } from '../../../../utils/sort.ts'

/**
 * Entry allowance shared by every operand of one `du` invocation.
 *
 * Backends with no native du op are walked one `readdir` at a time, which on an
 * API-backed tree is one request per directory. Slack, for instance, exposes a
 * directory per channel per day, so an unbounded walk of a real workspace is
 * tens of thousands of requests. The budget stops the walk and records that the
 * answer is partial.
 */
class WalkBudget {
  private remaining: number | null
  hit = false
  // Paths the walk was refused (a rule denied them below the operand), in
  // the order it met them; the generic reports them after the walks the
  // way GNU names an unreadable directory.
  readonly unreadable: string[] = []

  constructor(remaining: number | null) {
    this.remaining = remaining
  }

  /** Charge one entry; false once the cap is exhausted. */
  spend(): boolean {
    if (this.remaining === null) return true
    if (this.remaining <= 0) {
      this.hit = true
      return false
    }
    this.remaining -= 1
    return true
  }
}

/**
 * Account for an error the walk met at `path`, or rethrow it.
 *
 * Absence counts as zero, because an entry listed a moment ago can be
 * gone by the time the walk reaches it; `isMissingPath` is that set (a
 * stamped ENOENT, or a path outside every mount). A refusal counts as
 * zero too, but never silently: GNU `du` skips what it cannot read,
 * names it on stderr and exits 1, so the path is recorded here for the
 * generic to report. Everything else -- a 429, a 5xx, an aborted line --
 * says the walk never saw that subtree, and a confidently wrong size is
 * worse than an unknown one, so it surfaces instead of being summed as
 * nothing.
 *
 * Both of the walk's doors come through here, because a refused `stat` is
 * the same fact as a refused `readdir`: a rule denying a path outright
 * refuses before the walk ever learns the entry is a directory. Recording
 * only the `readdir` one left a refused subtree missing from the total
 * with du still exiting 0, which no caller can tell from a small tree.
 * The reported line is `cannot read directory` either way, which is the
 * wording GNU uses once it knows the entry is one; a refused *file* is
 * named by a noun that does not fit it, and being named loudly is still
 * the better half of that trade.
 */
function accountForWalkError(err: unknown, path: PathSpec, budget: WalkBudget): void {
  if (isEacces(err)) {
    budget.unreadable.push(path.virtual)
    return
  }
  if (!isMissingPath(err)) throw err
}

async function duWalk(
  ops: CommandIO,
  accessor: Accessor,
  index: IndexCacheStore | undefined,
  path: PathSpec,
  budget: WalkBudget,
  entries: [string, number][] | null,
): Promise<number> {
  let info
  try {
    info = await ops.stat(accessor, path, index)
  } catch (err) {
    accountForWalkError(err, path, budget)
    return 0
  }
  if (info.type !== FileType.DIRECTORY) {
    const size = info.size ?? 0
    if (entries !== null) {
      const prefix = mountPrefixOf(path.virtual, path.vfsPath)
      entries.push([`/${mountKey(path.virtual, prefix)}`, size])
    }
    return size
  }
  let children: string[]
  try {
    children = await ops.readdir(accessor, path, index)
  } catch (err) {
    accountForWalkError(err, path, budget)
    return 0
  }
  let total = 0
  for (const child of children) {
    if (!budget.spend()) break
    total += await duWalk(
      ops,
      accessor,
      index,
      PathSpec.fromStrPath(child, rekey(path.virtual, path.vfsPath, child)),
      budget,
      entries,
    )
  }
  return total
}

async function walkEntries(
  ops: CommandIO,
  accessor: Accessor,
  index: IndexCacheStore | undefined,
  budget: WalkBudget,
  path: PathSpec,
): Promise<DuEntries> {
  const entries: [string, number][] = []
  const total = await duWalk(ops, accessor, index, path, budget, entries)
  entries.sort((a, b) => compareCodePoints(a[0], b[0]))
  return [entries, total]
}

export const DU_BUILDER: Builder = {
  name: 'du',
  fn: async (ops, accessor, paths, _texts, opts) => {
    const idx = opts.index ?? undefined
    // A native du sums the raw tree; under a path rule the walk is what
    // reports a directory the rule refuses to open, where GNU does.
    const native = pathRulesActive() ? undefined : ops.du
    const budget = new WalkBudget(ops.maxDuEntries ?? DEFAULT_MAX_DU_ENTRIES)
    const computeSize: ComputeSize =
      native === undefined
        ? (p) => duWalk(ops, accessor, idx, p, budget, null)
        : (p) => native.size(accessor, p, idx)
    const computeEntries: ComputeEntries =
      native === undefined
        ? (p) => walkEntries(ops, accessor, idx, budget, p)
        : (p) => native.entries(accessor, p, idx)

    const out = await runDu(
      paths,
      opts,
      (targets) => resolveGlobOf(ops)(accessor, targets, idx),
      (p) => ops.stat(accessor, p, idx),
      computeSize,
      computeEntries,
      () => budget.hit,
      () => budget.unreadable,
    )
    return [out.stdout, new IOResult({ stderr: out.stderr, exitCode: out.exitCode })]
  },
}

File: python/mirage/commands/builtin/generic_bind/builders/du.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from dataclasses import dataclass
from functools import partial

from mirage.accessor.base import Accessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.generic.du import (ComputeEntries, ComputeSize,
                                                DuEntries, du_generic)
from mirage.commands.builtin.generic_bind.adapter import (Builder, CommandIO,
                                                          OperationFn)
from mirage.commands.config import CommandOpts
from mirage.io.types import ByteSource, IOResult
from mirage.types import FileType, PathSpec
from mirage.utils.key_prefix import mount_key, mount_prefix_of, rekey


@dataclass(slots=True)
class WalkBudget:
    """Entry allowance shared by every operand of one ``du`` invocation.

    Backends with no native du op are walked one ``readdir`` at a time,
    which on an API-backed tree is one request per directory. Slack, for
    instance, exposes a directory per channel per day, so an unbounded
    walk of a real workspace is tens of thousands of requests. The budget
    stops the walk and records that the answer is partial.

    Args:
        remaining (int | None): entries still allowed, or None for no cap.
        hit (bool): whether the cap was reached.
    """

    remaining: int | None
    hit: bool = False

    def spend(self) -> bool:
        """Charge one entry to the budget.

        Returns:
            bool: True if the walk may continue, False once exhausted.
        """
        if self.remaining is None:
            return True
        if self.remaining <= 0:
            self.hit = True
            return False
        self.remaining -= 1
        return True


async def _walk(
    ops: CommandIO,
    accessor: Accessor,
    index: IndexCacheStore,
    path: PathSpec,
    budget: WalkBudget,
    entries: list[tuple[str, int]] | None,
) -> int:
    try:
        info = await ops.stat(accessor, path, index)
    except (FileNotFoundError, ValueError):
        return 0
    if info.type != FileType.DIRECTORY:
        size = info.size or 0
        if entries is not None:
            prefix = mount_prefix_of(path.virtual, path.resource_path)
            entries.append(("/" + mount_key(path.virtual, prefix), size))
        return size
    try:
        children = await ops.readdir(accessor, path, index)
    except (FileNotFoundError, ValueError):
        return 0
    total = 0
    for child in children:
        if not budget.spend():
            break
        child_spec = PathSpec(virtual=child,
                              directory=child,
                              resolved=False,
                              resource_path=rekey(path.virtual,
                                                  path.resource_path, child))
        total += await _walk(ops, accessor, index, child_spec, budget, entries)
    return total


async def _walk_size(
    ops: CommandIO,
    accessor: Accessor,
    index: IndexCacheStore,
    budget: WalkBudget,
    path: PathSpec,
) -> int:
    return await _walk(ops, accessor, index, path, budget, None)


async def _walk_entries(
    ops: CommandIO,
    accessor: Accessor,
    index: IndexCacheStore,
    budget: WalkBudget,
    path: PathSpec,
) -> DuEntries:
    entries: list[tuple[str, int]] = []
    total = await _walk(ops, accessor, index, path, budget, entries)
    entries.sort()
    return entries, total


async def _op_size(
    op: OperationFn,
    accessor: Accessor,
    index: IndexCacheStore,
    path: PathSpec,
) -> int:
    return await op(accessor, path, index)


async def _op_entries(
    op: OperationFn,
    accessor: Accessor,
    index: IndexCacheStore,
    path: PathSpec,
) -> DuEntries:
    return await op(accessor, path, index)


async def _resolve(ops: CommandIO, accessor: Accessor, index: IndexCacheStore,
                   targets: list[PathSpec]) -> list[PathSpec]:
    return await ops.resolve_glob(accessor, targets, index)


async def _stat(ops: CommandIO, accessor: Accessor, index: IndexCacheStore,
                path: PathSpec):
    return await ops.stat(accessor, path, index)


def _budget_hit(budget: WalkBudget) -> bool:
    return budget.hit


async def du(ops: CommandIO, accessor: Accessor, paths: list[PathSpec],
             texts: list[str],
             opts: CommandOpts) -> tuple[ByteSource | None, IOResult]:
    if not ops.is_mounted(accessor):
        raise ValueError("du: no resource")
    budget = WalkBudget(ops.max_du_entries)
    native = ops.du
    compute_size: ComputeSize
    compute_entries: ComputeEntries
    if native is None:
        compute_size = partial(_walk_size, ops, accessor, opts.index, budget)
        compute_entries = partial(_walk_entries, ops, accessor, opts.index,
                                  budget)
    else:
        compute_size = partial(_op_size, native.size, accessor, opts.index)
        compute_entries = partial(_op_entries, native.entries, accessor,
                                  opts.index)
    return await du_generic(paths,
                            list(texts),
                            opts,
                            partial(_resolve, ops, accessor, opts.index),
                            partial(_stat, ops, accessor, opts.index),
                            compute_size,
                            compute_entries,
                            truncated=partial(_budget_hit, budget))


BUILDER = Builder('du', du, None, False, None)
